Fix: hurtling into wall of water

entrez opened the pull request:
A hero hurtled into a wall of water while levitating, flying, or wearing
water-walking boots wouldn't be stopped by it unless it was on the Plane
of Water.  Make it stop hurtling heroes immediately no matter the
location.

I also noticed that once I was hurtled safely into the wall of water, it
was described as a "pool" when I examined it with ':'.  Fix that, too,
even though I think it shouldn't really be encountered in-game.

entrez committed:
Fix: hurtling into wall of water

A hero hurtled into a wall of water while levitating, flying, or wearing
water-walking boots wouldn't be stopped by it unless it was on the Plane
of Water.  Make it stop hurtling heroes immediately no matter the
location.

I also noticed that once I was hurtled safely into the wall of water, it
was described as a "pool" when I examined it with ':'.  Fix that, too,
even though I think it shouldn't really be encountered in-game.

entrez commented:
I noticed this because of a Reddit post that Pat replied to a couple
days ago.  I figured a fix would be coming down the pipeline soon after
that comment, but since it hasn't yet, I'm submitting this thinking it
may help save a couple minutes finding and fixing it.

pat-rankin commented:
I started looking at that but got diverted.  One thing I noticed was that if I was levitating adjacent to a wall of water and threw an item away from it, I saw the animation of it traversing the throw's path, but after I recoiled into the water and escaped, I got the message "Splash!" and the thrown item ended up in the water instead of where it should have landed.  I haven't attempted to figure this one out but it is repeatable.  (It was just a priest throwing his wielded mace, not an aklys or anything else special.)

entrez commented:
Oh, that's interesting.  I didn't notice that because I was hurtling myself with #wiztelekinesis.  Looks like water_damage_chain is overwriting bhitpos, so when control returns to throwit after hurtle it thinks the item landed in the water location.  I guess you could probably fix it by caching bhitpos before the hurtle call and restoring it afterwards, but I haven't tested it yet.

entrez committed:
Cache and restore gb.bhitpos in water_damage_chain

If throwing an item while levitating sent the hero hurtling into a wall
of water, the item would land in the water due to water_damage_chain's
use of bhitpos.  Restore the previous value when it is finished to avoid
interfering with the use of bhitpos further up the call stack.

entrez commented:
I don't know if that's the right place to cache/restore it.  Maybe it should happen in water_damage_chain instead, since that seems to be the most direct culprit in this case.

entrez commented:
This is more directly connected with the overwriting on bhitpos and seems to work just as well as doing it in hurtle.

entrez commented:
This was merged as ccf997b and 5ba7939
